ql/src/java/org/apache/hadoop/hive/ql/parse/BaseSemanticAnalyzer.java

```diff
@@ -2047,8 +2070,8 @@ public void startAnalysis() {
   /**
    * Called when we end analysis of a query.
    */
-  public void endAnalysis() {
-    // Nothing to do
+  public void endAnalysis(ASTNode tree) {
+    setQueryType(tree); // at this point we know the query type for sure
```

Comment:
can't we set the queryType in `analyzeInternal(ast)`?

Reply:
we cannot, analyzeInternal is overriden by all the semantic analyzer children, and they are not supposed to be aware of this thing

Reply:
If the extra parameter is an issue maybe we can try doing one of these:

Option 1: Add a field `ASTNode tree` to `BaseSemanticAnalyzer`

Option 2: set it in `BaseSemanticAnalyzer.analyze`
```
  public void analyze(ASTNode ast, Context ctx) throws SemanticException {
    initCtx(ctx);
    init(true);
    analyzeInternal(ast);
    // New code: set query type after `analyzeInternal` completed
    setQueryType(ast).
  }
```
But in this case we have to do it in `ColumnStatsSemanticAnalyzer.java` too because it overrides `BaseSemanticAnalyzer.analyze`. However it always sets `QueryType.STATS`

Option 3: keep the extra parameter and leave this part as it is now.